File: libs/ultra-infer/python/ultra_infer/vision/classification/contrib/resnet.py
# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import absolute_import

from .... import UltraInferModel, ModelFormat
from .... import c_lib_wrap as C


class ResNet(UltraInferModel):
    def __init__(
        self,
        model_file,
        params_file="",
        runtime_option=None,
        model_format=ModelFormat.ONNX,
    ):
        """Load a image classification model exported by torchvision.ResNet.

        :param model_file: (str)Path of model file, e.g resnet/resnet50.onnx
        :param params_file: (str)Path of parameters file, if the model_fomat is ModelFormat.ONNX, this param will be ignored, can be set as empty string
        :param runtime_option: (ultra_infer.RuntimeOption)RuntimeOption for inference this model, if it's None, will use the default backend on CPU
        :param model_format: (ultra_infer.ModelForamt)Model format of the loaded model, default is ONNX
        """

        # call super() to initialize the backend_option
        # the result of initialization will be saved in self._runtime_option
        super(ResNet, self).__init__(runtime_option)

        self._model = C.vision.classification.ResNet(
            model_file, params_file, self._runtime_option, model_format
        )
        # self.initialized shows the initialization of the model is successful or not

        assert self.initialized, "ResNet initialize failed."

    # Predict and return the inference result of "input_image".
    def predict(self, input_image, topk=1):
        """Classify an input image

        :param input_image: (numpy.ndarray)The input image data, 3-D array with layout HWC, BGR format
        :param topk: (int)The topk result by the classify confidence score, default 1
        :return: ClassifyResult
        """
        return self._model.predict(input_image, topk)

    # Implement the setter and getter method for variables
    @property
    def size(self):
        """
        Returns the preprocess image size, default size = [224, 224];
        """
        return self._model.size

    @property
    def mean_vals(self):
        """
        Returns the mean value of normalization, default mean_vals = [0.485f, 0.456f, 0.406f];
        """
        return self._model.mean_vals

    @property
    def std_vals(self):
        """
        Returns the std value of normalization, default std_vals = [0.229f, 0.224f, 0.225f];
        """
        return self._model.std_vals

    @size.setter
    def size(self, wh):
        assert isinstance(
            wh, (list, tuple)
        ), "The value to set `size` must be type of tuple or list."
        assert (
            len(wh) == 2
        ), "The value to set `size` must contatins 2 elements means [width, height], but now it contains {} elements.".format(
            len(wh)
        )
        self._model.size = wh

    @mean_vals.setter
    def mean_vals(self, value):
        assert isinstance(
            value, list
        ), "The value to set `mean_vals` must be type of list."
        self._model.mean_vals = value

    @std_vals.setter
    def std_vals(self, value):
        assert isinstance(
            value, list
        ), "The value to set `std_vals` must be type of list."
        self._model.std_vals = value

File: deploy/ultra-infer/ultra_infer/vision/classification/contrib/resnet.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ultra_infer/vision/classification/contrib/resnet.h"
#include "ultra_infer/utils/perf.h"
#include "ultra_infer/vision/utils/utils.h"

namespace ultra_infer {
namespace vision {
namespace classification {

ResNet::ResNet(const std::string &model_file, const std::string &params_file,
               const RuntimeOption &custom_option,
               const ModelFormat &model_format) {
  // In constructor, the 3 steps below are necessary.
  // 1. set the Backend 2. set RuntimeOption 3. call Initialize()

  if (model_format == ModelFormat::ONNX) {
    valid_cpu_backends = {Backend::ORT, Backend::OPENVINO};
    valid_gpu_backends = {Backend::ORT, Backend::TRT};
  } else {
    valid_cpu_backends = {Backend::PDINFER};
    valid_gpu_backends = {Backend::PDINFER};
  }
  runtime_option = custom_option;
  runtime_option.model_format = model_format;
  runtime_option.model_file = model_file;
  runtime_option.params_file = params_file;
  initialized = Initialize();
}

bool ResNet::Initialize() {
  // In this function, the 3 steps below are necessary.
  // 1. assign values to the global variables 2. call InitRuntime()

  size = {224, 224};
  mean_vals = {0.485f, 0.456f, 0.406f};
  std_vals = {0.229f, 0.224f, 0.225f};

  if (!InitRuntime()) {
    FDERROR << "Failed to initialize ultra_infer backend." << std::endl;
    return false;
  }
  return true;
}

bool ResNet::Preprocess(Mat *mat, FDTensor *output) {
  // In this function, the preprocess need be implemented according to the
  // original Repos,
  // The result of preprocess has to be saved in FDTensor variable, because the
  // input of Infer() need to be std::vector<FDTensor>.
  // 1. Resize 2. BGR2RGB 3. Normalize 4. HWC2CHW 5. Put the result into
  // FDTensor variable.

  if (mat->Height() != size[0] || mat->Width() != size[1]) {
    int interp = cv::INTER_LINEAR;
    Resize::Run(mat, size[1], size[0], -1, -1, interp);
  }

  BGR2RGB::Run(mat);
  Normalize::Run(mat, mean_vals, std_vals);

  HWC2CHW::Run(mat);
  Cast::Run(mat, "float");
  mat->ShareWithTensor(output);
  output->shape.insert(output->shape.begin(), 1); // reshape to n, c, h, w
  return true;
}

bool ResNet::Postprocess(FDTensor &infer_result, ClassifyResult *result,
                         int topk) {
  // In this function, the postprocess need be implemented according to the
  // original Repos,
  // Finally the result of postprocess should be saved in ClassifyResult
  // variable.
  // 1. Softmax 2. Choose topk labels 3. Put the result into ClassifyResult
  // variable.

  int num_classes = infer_result.shape[1];
  function::Softmax(infer_result, &infer_result);
  const float *infer_result_buffer =
      reinterpret_cast<float *>(infer_result.Data());
  topk = std::min(num_classes, topk);
  result->label_ids =
      utils::TopKIndices(infer_result_buffer, num_classes, topk);
  result->scores.resize(topk);
  for (int i = 0; i < topk; ++i) {
    result->scores[i] = *(infer_result_buffer + result->label_ids[i]);
  }
  return true;
}

bool ResNet::Predict(cv::Mat *im, ClassifyResult *result, int topk) {
  // In this function, the Preprocess(), Infer(), and Postprocess() are called
  // sequentially.

  Mat mat(*im);
  std::vector<FDTensor> processed_data(1);
  if (!Preprocess(&mat, &(processed_data[0]))) {
    FDERROR << "Failed to preprocess input data while using model:"
            << ModelName() << "." << std::endl;
    return false;
  }
  processed_data[0].name = InputInfoOfRuntime(0).name;

  std::vector<FDTensor> output_tensors;
  if (!Infer(processed_data, &output_tensors)) {
    FDERROR << "Failed to inference while using model:" << ModelName() << "."
            << std::endl;
    return false;
  }

  if (!Postprocess(output_tensors[0], result, topk)) {
    FDERROR << "Failed to postprocess while using model:" << ModelName() << "."
            << std::endl;
    return false;
  }

  return true;
}

} // namespace classification
} // namespace vision
} // namespace ultra_infer
